### ISSUE TITLE
Implement A*

### ISSUE THREAD

**Issue by louis030195** [louis030195/diablo2-protocol#41]
Can already be useful with fixed layout maps such as acts camp

**Comment by rom1504** [comment]
https://www.npmjs.com/package/a-star

required : 
* a function (x,y) => wall or floor ; for the current location of the bot

if that's available, it's easy to use the package above

**Comment by rom1504** [comment]
now that walk verify is fixed, it's also possible to do something stupid : 
* to go from (xa,ya) to (xb,yb) do steps of like 100 of distance every time
* if walk verify says the bot didn't manage to get to one of the step, try to go right, try to go left, etc

**Comment by rom1504** [comment]
There are also algorithms for maze-solving https://en.wikipedia.org/wiki/Maze_solving_algorithm ie path finding without the map.
It could be interesting to see if that can be applied easily here.

**Comment by louis030195** [comment]
in durance of hate lvl 2
D2GS_REASSIGNPLAYER {"unitType":0,"unitId":1,"x":17714,"y":6801,"value":1}
D2GS_ASSIGNLVLWARP {"unitType":5,"unitId":3,"unknown":66,"x":17685,"y":8021,"warpId":104}

distance between player and warp = 1220
if we get all warps from that distance, easy path finding

**Comment by rom1504** [comment]
https://www.youtube.com/watch?v=6OzpKm4te-E

**Comment by rom1504** [comment]
http://www.primaryobjects.com/maze/ is a nice demo : 
* a* require to know the map because it takes several time at the same time => a bot cannot split itself in several
* tremaux algorithm (depth-first search) takes only one path at a time and backtrack if necessary => adapted to path finding without a map

**Comment by rom1504** [comment]
the implementation of tremaux algo is there https://github.com/primaryobjects/maze

**Comment by rom1504** [comment]
https://github.com/primaryobjects/maze/blob/master/scripts/tremauxAlgorithm.js

**Comment by rom1504** [comment]
https://en.wikipedia.org/wiki/Depth-first_search

**Comment by rom1504** [comment]
https://qiao.github.io/PathFinding.js/visual/

**Comment by rom1504** [comment]
https://justinsuen.github.io/visualgos/ : has dfs

**Comment by rom1504** [comment]
https://github.com/justinsuen/visualgos#depth-first-search

**Comment by rom1504** [comment]
blind search : http://www.cs.iusb.edu/~danav/teach/c463/4_blind_search.html

**Comment by rom1504** [comment]
-> https://www.geeksforgeeks.org/print-the-dfs-traversal-step-wise-backtracking-also/
that's what we want, dfs with backtracking

in our case we want to do the backtracking explicitely, the bot actually has to do it, it's not just theoritical.

**Comment by rom1504** [comment]
https://www.aaai.org/Papers/JAIR/Vol21/JAIR-2119.pdf

**Comment by rom1504** [comment]
Yes it has to go back to a previous unblocked location if it gets to a
blocked location

**Comment by rom1504** [comment]
If you don't have the map that's the only way to do it
